### PR TITLE
[FLINK-18562] Support for Hadoop ABFS  for Azure Datalake Gen2 accounts

### PR DESCRIPTION
**Pull request by srinipunuru**
## What is the purpose of the change

This pull request adds support for [abfs](https://hadoop.apache.org/docs/stable/hadoop-azure/abfs.html) to talk to ADLS Gen2 (Azure DataLake Gen2) storage accounts. These are newer storage account types that hierarchical namespaces, abfs takes advantage of this newer capability. Azure recommends using abfs for reading/writing to these newer storage accounts. 

## Brief change log
- Upgrade the hadoop version to latest. Existing version that flink depends on doesn't have hadoop abfs implementation. 
- Added support for two new schemes (abfs:// and abfss://)
- Added english documentation. 

## Verifying this change

- Ran existing tests to make sure there are no regressions
- Added few trivial unit tests. File system implementations are really hard to test using unit tests. Existing filesystems unit tests seems to be super thin. 
- Tested abfs, abfss manually using a flink clusters on AKS using the instructions [here](https://github.com/srinipunuru/flink-playground/blob/master/abfs-support/README.md)

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (**yes** / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (**yes** / no / don't know)
  - The S3 file system connector: (**yes** / no / don't know) - Because of the upgrade of the hadoop version

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / **docs** / JavaDocs / not documented) - Yes (English version)

### CODE REVIEW

**Comment by srinipunuru** [comment]
@flinkbot run azure

**Comment by srinipunuru** [comment]
Tagging @AHeise who is helping review this one.

File: flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AbstractAzureFSFactory.java [hunk 1]
```diff
@@ -75,18 +76,27 @@ public void configure(Configuration config) {
...
             URI fsUri, Configuration flinkConfig) throws IOException {
         org.apache.hadoop.conf.Configuration hadoopConfig = configLoader.getOrLoadHadoopConfig();
-
-        org.apache.hadoop.fs.FileSystem azureFS = new NativeAzureFileSystem();
-        azureFS.initialize(fsUri, hadoopConfig);
-
-        return azureFS;
+        String scheme = fsUri.getScheme();
+
+        if (scheme.startsWith("wasb")) {
```
Comment (AHeise) [inline flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AbstractAzureFSFactory.java:90]: How about adding a 
```abstract org.apache.hadoop.fs.FileSystem createAzureFS();```
method to this factory and avoid this if-else?
Then the abstract factory doesn't need to know the scheme of the implementations.

File: flink-filesystems/flink-fs-hadoop-shaded/src/main/resources/META-INF/NOTICE [hunk 2]
```diff
@@ -33,7 +48,7 @@ See bundled license files for details.
 
 - org.codehaus.woodstox:stax2-api:3.1.4 (https://github.com/FasterXML/stax2-api/tree/stax2-api-3.1.4)
 
-This project bundles org.apache.hadoop:*:3.1.0 from which it inherits the following notices:
+This project bundles org.apache.hadoop:*:3.3.1 from which it inherits the following notices:
 
 The Apache Hadoop project contains subcomponents with separate copyright
```
Comment (AHeise) [inline flink-filesystems/flink-fs-hadoop-shaded/src/main/resources/META-INF/NOTICE:53]: did you check if the notices changed?

File: flink-filesystems/flink-s3-fs-hadoop/src/main/java/org/apache/flink/fs/s3hadoop/HadoopS3AccessHelper.java [hunk 3]
```diff
@@ -52,8 +55,22 @@
     private final InternalWriteOperationHelper s3accessHelper;
 
     public HadoopS3AccessHelper(S3AFileSystem s3a, Configuration conf) {
```
Comment (AHeise) [inline flink-filesystems/flink-s3-fs-hadoop/src/main/java/org/apache/flink/fs/s3hadoop/HadoopS3AccessHelper.java:57]: The changes in this file are non-obvious to me: is it necessary because of the update or is this an additional feature? If it's the latter, please factor out into another commit.

File: flink-filesystems/flink-s3-fs-presto/src/main/resources/META-INF/NOTICE [hunk 4]
```diff
@@ -34,14 +34,25 @@ This project bundles the following dependencies under the Apache Software Licens
...
+- org.apache.httpcomponents:httpcore:4.4.14
+- org.apache.kerby:kerby-asn1:1.0.1
+- org.apache.kerby:kerb-core:1.0.1
+- org.apache.kerby:kerby-pkix:1.0.1
+- org.apache.kerby:kerby-util:1.0.1
+- org.codehaus.woodstox:stax2-api:4.2.1
+- org.xerial.snappy:snappy-java:1.1.8.3
 - org.weakref:jmxutils:1.19
+- org.wildfly.openssl:wildfly-openssl:1.0.7.Final
+- dnsjava:dnsjava:2.1.7
```
Comment (AHeise) [inline flink-filesystems/flink-s3-fs-presto/src/main/resources/META-INF/NOTICE:55]: Did you verify that this is indeed ASL?

**Review by AHeise — COMMENTED** [review]
Thank you very much for contributing! The code looks quite good already. I made a suggestion to disentangle the abstract factory from the specific factories.
Please double-check the licenses - two new modules are added; just to be safe.

Concerning testing your changes (and actually also the existing FS): there is now [azurite](https://github.com/Azure/Azurite) available which we could use inside a [testcontainer](https://www.testcontainers.org/) to have an ad-hoc blob storage. It would be a blast if you could add respective ITCases. I can give more detailed pointers or we create a follow-up task.

**Review by izzyacademy — COMMENTED** [review]
@srinipunuru I want to thank you very much for taking the time to work on this implementation. I am going to review the changes in more detail and I can work with you to update the Flink Documentation as soon as this PR is merged.

I will be reviewing your changes in this PR shortly.

Thanks again.

File: flink-filesystems/flink-s3-fs-presto/src/main/resources/META-INF/NOTICE (hunk 4 above)
Comment (izzyacademy) [reply flink-filesystems/flink-s3-fs-presto/src/main/resources/META-INF/NOTICE:55]: @AHeise after a quick check, these are the license for the two dependencies you have highlighted

DNSJava 2.1.7 is licensed under BSD-2 Clause
https://mvnrepository.com/artifact/dnsjava/dnsjava/2.1.7


WildFly OpenSSL is licensed under Apache 2.0 
https://mvnrepository.com/artifact/org.wildfly.openssl/wildfly-openssl/1.0.7.Final

/cc @srinipunuru

File: flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AbstractAzureFSFactory.java [hunk 5]
```diff
@@ -75,18 +76,27 @@ public void configure(Configuration config) {
...
+            LOG.info("Trying to initialize hadoop filesystem for {}.", scheme);
+            org.apache.hadoop.fs.FileSystem azureFS = new NativeAzureFileSystem();
+            azureFS.initialize(fsUri, hadoopConfig);
+            return azureFS;
+        } else {
```
Comment (izzyacademy) [inline flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AbstractAzureFSFactory.java:95]: Do we need to check if the scheme prefix starts with "abfs" here to cover both the secure and none secure schemes without including other schemes that are non Azure file systems?

File: flink-core/src/main/java/org/apache/flink/core/fs/FileSystem.java [hunk 6]
```diff
@@ -248,6 +248,8 @@
             ImmutableMultimap.<String, String>builder()
                     .put("wasb", "flink-fs-azure-hadoop")
                     .put("wasbs", "flink-fs-azure-hadoop")
+                    .put("abfs", "flink-fs-azure-hadoop")
+                    .put("abfss", "flink-fs-azure-hadoop")
```
Comment (izzyacademy) [inline flink-core/src/main/java/org/apache/flink/core/fs/FileSystem.java:252]: Thanks for adding these schemes in the Map. Otherwise, it will not allow those schemes to be recognized as supported FileSystems.

File: flink-filesystems/flink-s3-fs-presto/src/main/resources/META-INF/NOTICE (hunk 4 above)
Comment (AHeise) [reply flink-filesystems/flink-s3-fs-presto/src/main/resources/META-INF/NOTICE:55]: So `dnsjava` is definitively declared incorrectly then. Thanks for checking!

File: flink-core/src/main/java/org/apache/flink/core/fs/FileSystem.java (hunk 6 above)
Comment (AHeise) [reply flink-core/src/main/java/org/apache/flink/core/fs/FileSystem.java:252]: Just to clarify: Adding the entries will not change anything if everything is setup correctly. However, in case the user forgets to add the fs (as plugin or lib), this list can give a meaningful error message. So thanks for updating it!

**Comment by srinipunuru** [comment]
> Thank you very much for contributing! The code looks quite good already. I made a suggestion to disentangle the abstract factory from the specific factories.
> Please double-check the licenses - two new modules are added; just to be safe.
> 
> Concerning testing your changes (and actually also the existing FS): there is now [azurite](https://github.com/Azure/Azurite) available which we could use inside a [testcontainer](https://www.testcontainers.org/) to have an ad-hoc blob storage. It would be a blast if you could add respective ITCases. I can give more detailed pointers or we create a follow-up task.

Thanks for the Pointers @AHeise,  I did some digging last few days on the azurite. There is good news and bad news, Good news is, It looks like we could use that for testing with Azure storage with wasb:// (which is the legacy hdfs driver). It doesn't yet support the abfs and ADLS Gen2 https://github.com/Azure/Azurite/issues/553.  

I could add a follow up task to add a end to end test with Azurite for wasb. What do you think?

File: flink-filesystems/flink-s3-fs-presto/src/main/resources/META-INF/NOTICE (hunk 4 above)
Comment (srinipunuru) [reply flink-filesystems/flink-s3-fs-presto/src/main/resources/META-INF/NOTICE:55]: Thanks @AHeise , that's a miss. I will fix this and also verify the other ones as well.

File: flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AbstractAzureFSFactory.java (hunk 5 above)
Comment (srinipunuru) [reply flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AbstractAzureFSFactory.java:95]: Not necessarily, But anyway this code is going to get refactored and this if else condition will go away.

File: flink-filesystems/flink-fs-hadoop-shaded/src/main/resources/META-INF/NOTICE (hunk 2 above)
Comment (srinipunuru) [reply flink-filesystems/flink-fs-hadoop-shaded/src/main/resources/META-INF/NOTICE:53]: @AHeise How do i check if the notices have changed. Any pointers are appreciated. Thank you very much.

File: flink-filesystems/flink-s3-fs-hadoop/src/main/java/org/apache/flink/fs/s3hadoop/HadoopS3AccessHelper.java (hunk 3 above)
Comment (srinipunuru) [reply flink-filesystems/flink-s3-fs-hadoop/src/main/java/org/apache/flink/fs/s3hadoop/HadoopS3AccessHelper.java:57]: This change was needed because of the hadoop version upgrade.  Please let me know if this still requires a separate commit.

File: flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AbstractAzureFSFactory.java (hunk 1 above)
Comment (srinipunuru) [reply flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AbstractAzureFSFactory.java]: I have updated the PR, Please let me know if this looks OK. Thank you :)

File: flink-filesystems/flink-s3-fs-presto/src/main/resources/META-INF/NOTICE (hunk 4 above)
Comment (srinipunuru) [reply flink-filesystems/flink-s3-fs-presto/src/main/resources/META-INF/NOTICE]: Thanks for this comment, I did a scan of all the libraries, I fixed the licenses for all the libraries that are introduced because of the hadoop upgrade.

File: flink-filesystems/flink-s3-fs-hadoop/src/main/java/org/apache/flink/fs/s3hadoop/HadoopS3AccessHelper.java (hunk 3 above)
Comment (AHeise) [reply flink-filesystems/flink-s3-fs-hadoop/src/main/java/org/apache/flink/fs/s3hadoop/HadoopS3AccessHelper.java:57]: Yes, I think it would be best to have 2 commits ultimately.
In the first one, you are bumping the hadoop version and adjust this class + license/notice files. In the second commit, your actual changes come in.

**Comment by AHeise** [comment]
> Thanks for the Pointers @AHeise, I did some digging last few days on the azurite. There is good news and bad news, Good news is, It looks like we could use that for testing with Azure storage with wasb:// (which is the legacy hdfs driver). It doesn't yet support the abfs and ADLS Gen2 [Azure/Azurite#553](https://github.com/Azure/Azurite/issues/553).
> 
> I could add a follow up task to add a end to end test with Azurite for wasb. What do you think?
Thank you for investigating. I agree that a follow-up ticket is sufficient in this case as we cannot test the new features anyways. It would give us a bit more reassurance concerning the refactoring but I see it as an independent effort.

I suggest that once the commits are finished, you could do a final manual test with all 4 schemes and briefly mention it here.

File: flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AbstractAzureFSFactory.java (hunk 1 above)
Comment (AHeise) [reply flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AbstractAzureFSFactory.java]: From code structure, it now looks much cleaner. We could go one step further and pull the `initialize` back into this method. Then all `createAzureFS` are really just one liners.

File: flink-filesystems/flink-s3-fs-hadoop/src/main/java/org/apache/flink/fs/s3hadoop/HadoopS3AccessHelper.java (hunk 3 above)
Comment (srinipunuru) [reply flink-filesystems/flink-s3-fs-hadoop/src/main/java/org/apache/flink/fs/s3hadoop/HadoopS3AccessHelper.java:57]: That makes sense @AHeise, I updated the PR with just two commits as you described.

File: flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AbstractAzureFSFactory.java (hunk 1 above)
Comment (srinipunuru) [reply flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AbstractAzureFSFactory.java]: Great idea, That makes a lot of sense :).

File: flink-filesystems/flink-fs-hadoop-shaded/src/main/resources/META-INF/NOTICE (hunk 2 above)
Comment (AHeise) [reply flink-filesystems/flink-fs-hadoop-shaded/src/main/resources/META-INF/NOTICE:59]: I think you did a good job already. We also have some tooling on each CI run that does some checks.

**Comment by srinipunuru** [comment]
Thanks a lot for reviewing this @AHeise.  I updated the PR and I think i have addressed all of the comments. I verified the latest PR with all wasb/wasbss and abfs/abfss end to end. Please let me know if there is anything else pending to merge this PR. Thanks for your help.

File: flink-filesystems/flink-azure-fs-hadoop/src/main/resources/META-INF/services/org.apache.flink.core.fs.FileSystemFactory [hunk 7]
```diff
@@ -15,3 +15,5 @@
 
 org.apache.flink.fs.azurefs.AzureFSFactory
 org.apache.flink.fs.azurefs.SecureAzureFSFactory
+org.apache.flink.fs.azurefs.ABFSAzureFSFactory
```
Comment (izzyacademy) [inline flink-filesystems/flink-azure-fs-hadoop/src/main/resources/META-INF/services/org.apache.flink.core.fs.FileSystemFactory:18]: Later on, I would love for us to rename these factories to something that clarifies the filesystems they actually support.

For now, it is ok. It would just be a minor improvement for clarity @AHeise should we make that a FLIP?

- org.apache.flink.fs.azurefs.AzureFSFactory supports wasb scheme (Non-Secure Azure Blob Storage)
- org.apache.flink.fs.azurefs.SecureAzureFSFactory supports wasbs scheme (Secure Azure Blob Storage)
- org.apache.flink.fs.azurefs.ABFSAzureFSFactory supports abfs scheme (Non-Secure Azure Data Lake Store Gen 2)
- org.apache.flink.fs.azurefs.SecureABFSAzureFSFactory supports abfss scheme (Secure Azure Data Lake Store Gen 2)

We could rename them as follows:

- org.apache.flink.fs.azurefs.AzureFSFactory -> org.apache.flink.fs.azurefs.AzureBlobStorageFSFactory
- org.apache.flink.fs.azurefs.SecureAzureFSFactory -> org.apache.flink.fs.azurefs.SecureAzureBlobStorageFSFactory 
- org.apache.flink.fs.azurefs.ABFSAzureFSFactory -> org.apache.flink.fs.azurefs.AzureDataLakeStoreGen2FSFactory
- org.apache.flink.fs.azurefs.SecureABFSAzureFSFactory -> org.apache.flink.fs.azurefs.SecureAzureDataLakeStoreGen2FSFactory

What do you think @srinipunuru @AHeise ?

**Review by izzyacademy — APPROVED** [review]

File: docs/content/docs/deployment/filesystems/overview.md [hunk 8]
```diff
@@ -56,7 +56,7 @@ The Apache Flink project supports the following file systems:
   - **[Aliyun Object Storage Service]({{< ref "docs/deployment/filesystems/oss" >}})** is supported by `flink-oss-fs-hadoop` and registered under the *oss://* URI scheme.
   The implementation is based on the [Hadoop Project](https://hadoop.apache.org/) but is self-contained with no dependency footprint.
 
-  - **[Azure Blob Storage]({{< ref "docs/deployment/filesystems/azure" >}})** is supported by `flink-azure-fs-hadoop` and registered under the *wasb(s)://* URI schemes.
+  - **[Azure Blob Storage]({{< ref "docs/deployment/filesystems/azure" >}})** is supported by `flink-azure-fs-hadoop` and registered under the *abfs(s)://* and *wasb(s)://* URI schemes.
```
[the remark this answers is not in the document]
Comment (AHeise) [reply docs/content/docs/deployment/filesystems/overview.md:59]: I like this idea and would like to see it prior to merging.

File: docs/content/docs/deployment/filesystems/azure.md [hunk 9]
```diff
@@ -83,4 +103,19 @@ environment variable `AZURE_STORAGE_KEY` by setting the following configuration
 fs.azure.account.keyprovider.<account_name>.blob.core.windows.net: org.apache.flink.fs.azurefs.EnvironmentVariableKeyProvider
 ```
 
+### ABFS
+
+Hadoop's ABFS Azure Filesystem supports several ways of configuring authentication. Please visit the [Hadoop ABFS documentation](https://hadoop.apache.org/docs/stable/hadoop-azure/abfs.html#Authentication) documentation on how to configure.
+
+{{< hint info >}}
+Azure recommends using Azure managed identity to access the ADLS Gen2 storage accounts using abfs. Details on how to do this are beyond the scope of this documentation, please refer to the Azure documentation for more details.
+{{< /hint >}}
```
[the remark this answers is not in the document]
Comment (AHeise) [reply docs/content/docs/deployment/filesystems/azure.md:112]: Anything that helps the user to setup the filesystem correctly without trial and error is always good. However, I'd be completely fine in doing that in a follow-up PR; for me, it doesn't seem strongly connected to this change (but I could be wrong).

File: flink-filesystems/flink-azure-fs-hadoop/src/main/resources/META-INF/services/org.apache.flink.core.fs.FileSystemFactory (hunk 7 above)
Comment (AHeise) [reply flink-filesystems/flink-azure-fs-hadoop/src/main/resources/META-INF/services/org.apache.flink.core.fs.FileSystemFactory:18]: It's not user-facing, so a FLIP is not needed at all. In fact, I'm assuming that it would be best to do it immediately here. Rename the old factories in 1 additional commit in this PR and rename the new factories in the main commit of this PR.

**Comment by AHeise** [comment]
> Thanks a lot for reviewing this @AHeise. I updated the PR and I think i have addressed all of the comments. I verified the latest PR with all wasb/wasbss and abfs/abfss end to end. Please let me know if there is anything else pending to merge this PR. Thanks for your help.

Thank you again for your contribution and the quick responses. Except for the smaller suggestion of @izzyacademy , everything looks good. Could you please prefix all commits with [FLINK-18562][fs]? Then it's easier to see what the commits are about when they are merged into master.

File: docs/content/docs/deployment/filesystems/azure.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (srinipunuru) [reply docs/content/docs/deployment/filesystems/azure.md]: Thanks for the links @izzyacademy Instead of adding the list of Azure services directly in the Flink documentation which may get outdated, I added links to the page. I hope this looks ok.

File: flink-filesystems/flink-azure-fs-hadoop/src/main/resources/META-INF/services/org.apache.flink.core.fs.FileSystemFactory (hunk 7 above)
Comment (srinipunuru) [reply flink-filesystems/flink-azure-fs-hadoop/src/main/resources/META-INF/services/org.apache.flink.core.fs.FileSystemFactory]: Good feedback. I updated the file names in the latest change.

**Comment by srinipunuru** [comment]
> > Thanks a lot for reviewing this @AHeise. I updated the PR and I think i have addressed all of the comments. I verified the latest PR with all wasb/wasbss and abfs/abfss end to end. Please let me know if there is anything else pending to merge this PR. Thanks for your help.
> 
> Thank you again for your contribution and the quick responses. Except for the smaller suggestion of @izzyacademy , everything looks good. Could you please prefix all commits with [[FLINK-18562](https://issues.apache.org/jira/browse/FLINK-18562)][fs]? Then it's easier to see what the commits are about when they are merged into master.

Thanks @AHeise I have addressed the comments and also updated the commit messages according to the convention followed. Thanks again for taking time to review this :)

File: docs/content/docs/deployment/filesystems/overview.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (srinipunuru) [reply docs/content/docs/deployment/filesystems/overview.md]: Great idea, I made updates to differentiate the ADLS Gen2 and Azure Blob storage.

**Review by AHeise — APPROVED** [review]
LGTM. Thank you very much for the contribution. I'm merging now.
